src/Development/Rattle/Profile.hs

```diff
@@ -0,0 +1,252 @@
+  cmds <- maybe (return []) (getSpeculate shared) rattleSpeculate
+  fmap (takeWhile (not . List.null . snd)) $ forM cmds $ \x -> (x,) <$> unsafeInterleaveIO (getCmdTraces shared x)
+  
+constructGraph :: RattleOptions -> IO Graph
+constructGraph options@RattleOptions{..} = do
+  cmdsWTraces <- getCmdsTraces options
+  return $ createGraph cmdsWTraces
+
+graphData :: RattleOptions -> IO (Seconds,Seconds,Seconds)
+graphData options = do
+  cmdsWTraces <- getCmdsTraces options
+  let graph = createGraph cmdsWTraces in
+    let w = work graph
+        s = spanGraph graph in
+      return (w,s,(w / s))
+
+writeProfile :: RattleOptions -> FilePath -> IO ()
+writeProfile options out = do
+  graph <- constructGraph options
+  writeProfileInternal out graph
+
+writeProfileInternal :: FilePath -> Graph -> IO ()
+writeProfileInternal out g = LBS.writeFile out =<< generateHTML g
+
+{- build graph using trace info
+   Add an edge between 2 nodes if they both write the same file
+   Add an edge between 2 nodes if one reads a file and another writes it -}
+createGraph :: [(Cmd,[Trace Hash])] -> Graph
+createGraph xs = Graph xs $ g xs
+  where g [] = []
+        g (x:xs) = let edges = Maybe.mapMaybe (createEdge x) xs in
+          edges ++ (g xs)  
+
+-- assume p1 occurred before p2. 
+-- find the worst type of hazard if there is an edge
+createEdge :: (Cmd,[Trace Hash]) -> (Cmd,[Trace Hash]) -> Maybe Edge
+createEdge p1@(cmd1,ts) p2@(cmd2,ls) = -- first look for write write hazard then look for both read/write and write/read hazards
+  case writeWriteHazard ts ls of
+    Just fp -> Just $ Edge p1 p2 $ Just $ WriteWriteHazard fp cmd1 cmd2
+    Nothing -> case readWriteHazard ts ls of
+                 Just fp -> Just $ Edge p1 p2 $ Just $ ReadWriteHazard fp cmd1 cmd2
+                 Nothing -> -- check for a non hazard edge
+                   case readWriteHazard ls ts of
+                     Just fp -> Just $ Edge p1 p2 Nothing -- regular edge
+                     Nothing -> Nothing -- no edge
+    
+-- Is there a writewrite edge?
+writeWriteHazard :: [Trace Hash] -> [Trace Hash] -> Maybe FilePath
+writeWriteHazard = maybeHazard tWrite 
+
+-- Is there a readwrite edge?
+readWriteHazard :: [Trace Hash] -> [Trace Hash] -> Maybe FilePath
+readWriteHazard = maybeHazard tRead
+
+maybeHazard :: ((Trace Hash) -> [(FilePath, Hash)]) -> [Trace Hash] -> [Trace Hash] -> Maybe FilePath
+maybeHazard _ [] ls = Nothing
+maybeHazard _ ls [] = Nothing
+maybeHazard f (t:ts) ls =
+  case find (\y -> case memberWrites y ls of
+                     Just _ -> True
+                     Nothing -> False) $ f t of
+    Just (fp,_) -> Just fp
+    Nothing -> maybeHazard f ts ls
+
+memberWrites :: (FilePath, Hash) -> [Trace Hash] -> Maybe FilePath
+memberWrites x [] = Nothing
+memberWrites x@(fp,_) (y:ys) =
+  case find (\(fp2,a) -> fp == fp2) $ tWrite y of
+    Just (fp,_) -> Just fp
+    Nothing -> memberWrites x ys
```

Comment:
`lookup fp $ tWrite y` ?

Reply:
I think that would return a rather than (fp2,a).

Reply:
`fp == fp2` so `fmap (fp,)` around lookup would work.